azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/sync_clients.py

```diff
@@ -151,6 +154,52 @@ def callback():
         self._transport.send_event(message, callback=callback)
         send_complete.wait()
 
+    def receive_method(self, method_name=None, block=True, timeout=None):
+        """Receive a method call via the Azure IoT Hub or Azure IoT Edge Hub.
+
+        :param str method_name: Optionally provide the name of the method to receive calls for.
+        If this parameter is not given, all methods not already being specifically targeted by
+        a different call to receive_method will be received.
+        :param bool block: Indicates if the operation should block until a message is received.
+        Default True.
+        :param int timeout: Optionally provide a number of seconds until blocking times out.
+
+        :raises: InboxEmpty if timeout occurs on a blocking operation.
+        :raises: InboxEmpty if no message is available on a non-blocking operation.
+
+        :returns: MethodCall object representing the received method call.
+        """
+        if not self._transport.feature_enabled[constant.METHODS]:
+            self._enable_feature(constant.METHODS)
+
+        if method_name:
+            method_inbox = self._inbox_manager.get_named_method_call_inbox(method_name)
+        else:
+            method_inbox = self._inbox_manager.get_generic_method_call_inbox()
+
+        logger.info("Waiting for method call...")
+        method_call = method_inbox.get(block=block, timeout=timeout)
+        logger.info("Received method call")
+        return method_call
+
+    def send_method_response(self, method, result, status):
```

Comment:
>send_method_response [](start = 8, length = 20)

question: the success of methods depends on the device staying connected while it is processed by the user. if we get disconnected with stuff in the method inbox, how would we detect that and what should we do? probably clean up so that the user does not get notified of methods that are doomed to fail?
And then what happens if the user tries to respond to a method that has already failed because the device has been disconnected?

Reply:
I'm unclear what you mean by this. My understanding was as long as you've got a request ID, and the timeout hasn't occurred, that you'd be able to respond, as long as you had reconnected.  Can you clarify why items hanging in the inbox would be "doomed to fail"? If it's a question of timeouts, we ask the user to verify that the timeout hasn't already occurred.

---
In reply to: [266186191](https://github.com/Azure/azure-iot-sdk-python-preview/pull/61#discussion_r266186191) [](ancestors = 266186191)

Reply:
your understanding is incorrect. if a device is disconnected, methods immediately fail on the service side. if you reconnect and call with an old request id, it won't work.

---
In reply to: [266627364](https://github.com/Azure/azure-iot-sdk-python-preview/pull/61#discussion_r266627364) [](ancestors = 266627364,266186191)